packages/gensx/tests/checkpoint.test.tsx

```diff
@@ -0,0 +1,359 @@
+  });
+
+  test("handles sequential execute calls within component", async () => {
+    // Define a simple component that we'll use multiple times
+    const SimpleComponent = gsx.Component<{ id: number }, string>(
+      "SimpleComponent",
+      async ({ id }) => {
+        await setTimeout(0);
+        return `component ${id}`;
+      },
+    );
+
+    // Create a component that makes three sequential execute calls
+    const ParentComponent = gsx.Component<{}, string>(
+      "ParentComponent",
+      async () => {
+        const first = await gsx.execute<string>(<SimpleComponent id={1} />);
+        const second = await gsx.execute<string>(<SimpleComponent id={2} />);
+        const third = await gsx.execute<string>(<SimpleComponent id={3} />);
+        return `${first}, ${second}, ${third}`;
+      },
+    );
+
+    // Execute workflow and get results
+    const { result, checkpoints } = await executeWithCheckpoints<string>(
+      <ParentComponent />,
+    );
+
+    // Verify execution result
+    expect(result).toBe("component 1, component 2, component 3");
+
+    // Get final checkpoint state
+    const finalCheckpoint = checkpoints[checkpoints.length - 1];
+    expect(finalCheckpoint.componentName).toBe("ParentComponent");
+
+    // For now, verify we have the right number of total nodes (1 parent + 3 children)
+    function countNodes(node: ExecutionNode): number {
+      return (
+        1 + node.children.reduce((sum, child) => sum + countNodes(child), 0)
+      );
+    }
+    expect(countNodes(finalCheckpoint)).toBe(4);
+    expect(finalCheckpoint.children.length).toBe(3);
+  });
+
+  test("handles component children with object return", async () => {
+    // Define child components
+    const ComponentA = gsx.Component<{ value: string }, string>(
+      "ComponentA",
+      async ({ value }) => {
+        await setTimeout(0);
+        return `a:${value}`;
+      },
+    );
+
+    const ComponentB = gsx.Component<{ value: string }, string>(
+      "ComponentB",
+      async ({ value }) => {
+        await setTimeout(0);
+        return `b:${value}`;
+      },
+    );
+
+    // Create parent component that returns an object with multiple children
+    const ParentComponent = gsx.Component<{}, { a: string; b: string }>(
+      "ParentComponent",
+      () => {
+        return {
+          a: <ComponentA value="first" />,
+          b: <ComponentB value="second" />,
+        };
+      },
+    );
+
+    // Execute workflow and get results
+    const { result, checkpoints } = await executeWithCheckpoints<{
+      a: string;
+      b: string;
+    }>(<ParentComponent />);
+
+    // Verify execution result
+    expect(result).toEqual({
+      a: "a:first",
+      b: "b:second",
+    });
+
+    // Get final checkpoint state
+    const finalCheckpoint = checkpoints[checkpoints.length - 1];
+
+    // Verify checkpoint structure
+    expect(finalCheckpoint).toMatchObject({
+      componentName: "ParentComponent",
```

Comment:
Do you think we are losing valuable information without including the output shape?

An alternative result here could be:

```
{
  comonentName: "ParentComponent",
  output: {
    a: {
      componentName: "ComponentA",
      ...
    },
    b: {
      componentName: "ComponentB",
      ...
    }
  },
}
```

Reply:
updated.